File: objectiveai-js/src/functions/executions/response/unary/function_execution.ts
import z from "zod";
import { TaskSchema } from "./task";
import { ReasoningSummarySchema } from "./reasoning_summary";
import { JsonValueSchema } from "src/json";
import { ObjectiveAIErrorSchema } from "src/error";
import { UsageSchema } from "src/vector/completions/response/usage";
import { ResponseObjectSchema } from "./response_object";

export const FunctionExecutionSchema = z
  .object({
    id: z.string().describe("The unique identifier of the function execution."),
    tasks: z
      .array(TaskSchema)
      .describe("The tasks executed as part of the function execution."),
    tasks_errors: z
      .boolean()
      .describe(
        "When true, indicates that one or more tasks encountered errors during execution."
      ),
    reasoning: ReasoningSummarySchema.nullable(),
    output: z
      .union([
        z.number().describe("The scalar output of the function execution."),
        z
          .array(z.number())
          .describe("The vector output of the function execution."),
        JsonValueSchema.describe(
          "The erroneous output of the function execution."
        ),
      ])
      .describe("The output of the function execution."),
    error: ObjectiveAIErrorSchema.nullable().describe(
      "When non-null, indicates that an error occurred during the function execution."
    ),
    retry_token: z
      .string()
      .nullable()
      .describe("A token which may be used to retry the function execution."),
    created: z
      .uint32()
      .describe(
        "The UNIX timestamp (in seconds) when the function execution chunk was created."
      ),
    function: z
      .string()
      .nullable()
      .describe("The unique identifier of the function being executed."),
    profile: z
      .string()
      .nullable()
      .describe("The unique identifier of the profile being used."),
    object: ResponseObjectSchema,
    usage: UsageSchema,
  })
  .describe("A function execution.");
export type FunctionExecution = z.infer<typeof FunctionExecutionSchema>;

File: objectiveai/objectiveai/src/functions/executions/response/unary/function_execution.rs
//! Complete function execution response.

use crate::{
    error,
    functions::{self, executions::response},
    vector,
};
use serde::{Deserialize, Serialize};

/// A complete function execution response (non-streaming).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FunctionExecution {
    /// Unique identifier for this execution.
    pub id: String,
    /// Results from each task in the function.
    pub tasks: Vec<super::Task>,
    /// Whether any tasks encountered errors.
    pub tasks_errors: bool,
    /// Reasoning summary if reasoning was enabled.
    pub reasoning: Option<super::ReasoningSummary>,
    /// The final output (scalar or vector score).
    pub output: functions::expression::FunctionOutput,
    /// Error details if the execution failed.
    pub error: Option<error::ResponseError>,
    /// Token for retrying this execution with cached votes.
    pub retry_token: Option<String>,
    /// Unix timestamp when the execution was created.
    pub created: u64,
    /// ID of the function used (if remote).
    pub function: Option<String>,
    /// ID of the profile used (if remote).
    pub profile: Option<String>,
    /// Object type identifier.
    pub object: super::Object,
    /// Aggregated token and cost usage.
    pub usage: vector::completions::response::Usage,
}

impl FunctionExecution {
    pub fn any_usage(&self) -> bool {
        self.usage.any_usage()
    }
}

impl From<response::streaming::FunctionExecutionChunk> for FunctionExecution {
    fn from(
        response::streaming::FunctionExecutionChunk {
            id,
            tasks,
            tasks_errors,
            reasoning,
            output,
            error,
            retry_token,
            created,
            function,
            profile,
            object,
            usage,
        }: response::streaming::FunctionExecutionChunk,
    ) -> Self {
        Self {
            id,
            tasks: tasks.into_iter().map(super::Task::from).collect(),
            tasks_errors: tasks_errors.unwrap_or(false),
            reasoning: reasoning.map(super::ReasoningSummary::from),
            output: output.unwrap_or(
                functions::expression::FunctionOutput::Err(
                    serde_json::Value::Null,
                ),
            ),
            error,
            retry_token,
            created,
            function,
            profile,
            object: object.into(),
            usage: usage.unwrap_or_default(),
        }
    }
}
